src/compute-client/src/controller/instance.rs

```diff
@@ -1559,226 +1447,85 @@ where
-                Some(read_policy) => {
-                    // For readable collections the read frontier is determined by applying the
-                    // client-provided read policy to the write frontier.
-                    read_policy.frontier(new_upper.borrow())
-                }
-                None => {
-                    // Write-only collections cannot be read within the context of the compute
-                    // controller, so we can immediately advance their read frontier to the new write
-                    // frontier.
-                    new_upper.clone()
-                }
-            };
+        // Report the frontier advancement internally and externally.
+        collection
+            .collection_introspection
+            .frontier_update(&new_frontier);
 
-            if PartialOrder::less_than(old_since, &new_since) {
-                let mut update = ChangeBatch::new();
-                update.extend(old_since.iter().map(|t| (t.clone(), -1)));
-                update.extend(new_since.iter().map(|t| (t.clone(), 1)));
-                read_capability_changes.insert(*id, update);
-                collection.implied_capability = new_since;
-            }
-        }
-        if !read_capability_changes.is_empty() {
-            self.update_read_capabilities(read_capability_changes);
-        }
+        self.deliver_response(ComputeControllerResponse::FrontierUpper {
+            id,
+            upper: new_frontier,
+        });
     }
 
-    /// Applies `updates`, propagates consequences through other read capabilities, and sends
-    /// appropriate compaction commands.
-    #[mz_ore::instrument(level = "debug")]
-    fn update_read_capabilities(&mut self, mut updates: BTreeMap<GlobalId, ChangeBatch<T>>) {
-        // Records storage read capability updates.
-        let mut storage_updates = BTreeMap::default();
-        // Records compute collections with downgraded read frontiers.
-        let mut compute_downgraded = BTreeSet::default();
-
-        // We must not rely on any specific relative ordering of `GlobalId`s.
-        // That said, it is reasonable to assume that collections generally have greater IDs than
-        // their dependencies, so starting with the largest is a useful optimization.
-        while let Some((id, mut update)) = updates.pop_last() {
+    /// Apply collection read hold changes pending in `read_holds_rx`.
+    fn apply_read_hold_changes(&mut self) {
+        let mut allowed_compaction = BTreeMap::new();
+
+        while let Ok((id, mut update)) = self.read_holds_rx.try_recv() {
             let Some(collection) = self.collections.get_mut(&id) else {
-                tracing::error!(
-                    %id, ?update,
-                    "received read capability update for an unknown collection",
+                soft_panic_or_log!(
+                    "read hold change for absent collection (id={id}, changes={update:?})"
```

Comment:
Now that updates come through a channel what guarantees that this won't produce false positives where a collection is dropped and then later we process the downgrades? Is the cleanup logic actually in this function (and therefore it won't run until we observe an empty read frontier)?

Reply:
We only remove collection state once there are no read capabilities on that collection anymore: https://github.com/teskje/materialize/blob/406a4c827190b40dbd65af1712eaf999362cb138/src/compute-client/src/controller/instance.rs#L660. If there are no read capabilities anymore, seeing updates would be a bug, as updates may only clone or downgrade existing capabilities.